engineering/frontend/tech-ecosystem.md

```diff
@@ -0,0 +1,95 @@
+# Tech ecosystem
+
+This document is intended to give you a taste of how we build frontend applications at Dwarves, and more importantly, why we choose the technology. It still assumes basic knowledge of React and [Next.js](https://nextjs.org/docs/getting-started). **If you're completely new to React, please refer to https://github.com/petehunt/react-howto instead!**
+
+## Tech stack
+
+### Core
+
+- [x] [React](https://reactjs.org/)
+- [x] [Next.js](https://nextjs.org/)
+- [x] [SWR](https://swr.vercel.app/)
+- [x] [React Hook Form](https://github.com/react-hook-form/react-hook-form)
+- [x] [Dwarves React Toolkit](https://github.com/dwarvesf/react-toolkit)
+
+### UI & styling
+
+- [x] [Tailwindcss](https://github.com/tailwindcss/tailwindcss)
+- [x] [Headless UI](https://github.com/tailwindlabs/headlessui)
+
+### Static type checking & linting
+
+- [x] [TypeScript](https://www.typescriptlang.org)
+- [x] [ESLint](http://eslint.org/)
+- [x] [Prettier](https://prettier.io/)
+
+### Testing
+
+- [x] [Jest](http://facebook.github.io/jest/)
+- [x] [react-testing-library](https://github.com/kentcdodds/react-testing-library)
+- [x] [Cypress](https://github.com/cypress-io/cypress)
+
+## Basic Building Blocks
+
+### Next.js
+
+Production grade React applications that scale, we list [Next.js](https://nextjs.org/) to be the first candidate when we decide to setup a React app. The opinionated framework gives you the best developer experience with all the features you need for production: hybrid static & server rendering, TypeScript support, smart bundling, route pre-fetching.
+
+### TypeScript
+
+We are adopting TypeScript to our codebase. Forcing to write strongly-typed syntax while you are familiar with dynamic/multi-paradigm scripting language like JavaScript is sometimes annoying, even for experienced developers. However, the drawback can be outweighed by the following benefits when coming to write a medium to large sized applications:
+
+- [x] Catching your bugs at compile-time as you type them.
+- [x] Coding confidently with autocompletion, definition jumping and source documentation.
+- [x] Syncing API Interfaces between backend and frontend via Swagger json doc.
+- [x] Relieve the pain of refactoring/renaming.
+
+### SWR
+
+We use [SWR](https://swr.vercel.app/) to be the main way of retrieving data in React apps. The strategy is to first return the data from cache (stale), then send the fetch request (revalidate), and finally, come with the up-to-date data. With SWR, components will get a stream of data updates constantly and automatically. And the UI will be always fast and reactive. It features:
+
+- [x] Transport and protocol agnostic data fetching
+- [x] Fast page navigation
+- [x] Revalidation on focus
+- [x] Interval polling
+- [x] Request deduplication
+- [x] Local mutation
+
```

Comment:
I think we should include fetch libs like `node-fetch` or `axios` or `unfetch` since swr is cache layer it required a fetcher

Reply:
Yes, the description seems misleading. I didn't aim to recommend a fetching library in the docs.

Reply:
Updated

> ### SWR
> We use [SWR](https://swr.vercel.app/) as the caching layer for the backend data to ensure the UI will be always fast and reactive. The strategy is to first return the data from cache (stale), then send the fetch request (revalidate), and finally, come with the up-to-date data. The data can be dynamic and self-updated overtime.